test.js

```diff
@@ -0,0 +1,252 @@
+        '0x000000000000000000000000000000000000dEa2': '45662'
+      })
+    }
+    {
+      const res = await fetch(`${api}/log`)
+      const log = await res.json()
+      for (const l of log) {
+        assert(l.timestamp)
+        delete l.timestamp
+      }
+      assert.deepEqual(log, [
+        {
+          address: '0x000000000000000000000000000000000000dEaD',
+          score: '10',
+          scheduledRewards: '4566'
+        },
+        {
+          address: '0x000000000000000000000000000000000000dEa2',
+          score: '100',
+          scheduledRewards: '45662'
+        },
+        {
+          address: '0x000000000000000000000000000000000000dEaD',
+          score: '10',
+          scheduledRewards: '4566'
+        }
+      ])
+    }
+  })
+  await t.test('paid rewards', async t => {
+    {
+      const rewards = {
+        '0x000000000000000000000000000000000000dEaD': '9132'
+      }
+      const res = await fetch(`${api}/paid`, {
+        method: 'POST',
+        body: JSON.stringify({
+          rewards,
+          signature: await sign(rewards)
+        })
+      })
+      assert.strictEqual(res.status, 200)
+      assert.deepStrictEqual(await res.json(), {
+        '0x000000000000000000000000000000000000dEaD': '0'
+      })
+    }
+    {
+      const res = await fetch(`${api}/scheduled-rewards`)
+      assert.deepEqual(await res.json(), {
+        '0x000000000000000000000000000000000000dEaD': '0',
+        '0x000000000000000000000000000000000000dEa2': '45662'
+      })
+    }
+    {
+      const res = await fetch(`${api}/log`)
+      const log = await res.json()
+      for (const l of log) {
+        assert(l.timestamp)
+        delete l.timestamp
+      }
+      assert.deepEqual(log, [
+        {
+          address: '0x000000000000000000000000000000000000dEaD',
+          score: '10',
+          scheduledRewards: '4566'
+        },
+        {
+          address: '0x000000000000000000000000000000000000dEa2',
+          score: '100',
+          scheduledRewards: '45662'
+        },
+        {
+          address: '0x000000000000000000000000000000000000dEaD',
+          score: '10',
+          scheduledRewards: '4566'
+        },
+        {
+          address: '0x000000000000000000000000000000000000dEaD',
+          scheduledRewards: '-9132'
+        }
+      ])
+    }
+  })
+  await t.test('validate signatures', async t => {
+    const digest = ethers.solidityPackedKeccak256(
+      ['address[]', 'uint256[]'],
+      [['0x000000000000000000000000000000000000dEaD'], ['2']]
+    )
+    const signed = await signer.signMessage(digest)
+    const { v, r, s } = ethers.Signature.from(signed)
+    const res = await fetch(`${api}/scores`, {
+      method: 'POST',
+      body: JSON.stringify({
+        scores: {
+          '0x000000000000000000000000000000000000dEaD': '1'
+        },
+        signature: {
+          v,
+          r,
+          s
+        }
+      })
+    })
+    assert.strictEqual(res.status, 403)
+  })
```

Comment:
Please also add a test to verify that the service rejects messages signed by a wallet that's not white-listed.

Reply:
done!